File: src/typescript-generator/src/mapping/validatorMapping.ts
import { ValidationRule } from '../models/validationRuleModels';

/**
 * Maps JSON validator names to TypeScript fluentvalidation-ts method calls
 */
export class ValidatorMapping {
  /**
   * Generates the TypeScript fluentvalidation-ts method call for a given validator and its parameters
   */
  static mapToTypeScript(validatorName: string, parameters?: Record<string, any>): string {
    switch (validatorName) {
      case 'NotNull':
        return 'notNull()';
      case 'NotEmpty':
        return 'notEmpty()';
      case 'Empty':
        return 'empty()';

      case 'Equal':
        return this.generateEqual(parameters);
      case 'NotEqual':
        return this.generateNotEqual(parameters);

      case 'Length':
        return this.generateLength(parameters);
      case 'MinLength':
        return this.generateMinLength(parameters);
      case 'MaxLength':
        return this.generateMaxLength(parameters);

      case 'EmailAddress':
        return 'emailAddress()';

      case 'Matches':
        return this.generateMatches(parameters);

      case 'LessThan':
        return this.generateLessThan(parameters);
      case 'LessThanOrEqualTo':
        return this.generateLessThanOrEqualTo(parameters);
      case 'GreaterThan':
        return this.generateGreaterThan(parameters);
      case 'GreaterThanOrEqualTo':
        return this.generateGreaterThanOrEqualTo(parameters);

      case 'InclusiveBetween':
        return this.generateInclusiveBetween(parameters);
      case 'ExclusiveBetween':
        return this.generateExclusiveBetween(parameters);

      default:
        throw new Error(`Validator '${validatorName}' is not supported`);
    }
  }

  private static generateEqual(parameters?: Record<string, any>): string {
    const value = this.getRequiredParameter(parameters, 'value', 'Equal');
    return `equal(${this.formatValue(value)})`;
  }

  private static generateNotEqual(parameters?: Record<string, any>): string {
    const value = this.getRequiredParameter(parameters, 'value', 'NotEqual');
    return `notEqual(${this.formatValue(value)})`;
  }

  private static generateLength(parameters?: Record<string, any>): string {
    const min = this.getRequiredParameter(parameters, 'min', 'Length');
    const max = this.getRequiredParameter(parameters, 'max', 'Length');
    return `length(${min}, ${max})`;
  }

  private static generateMinLength(parameters?: Record<string, any>): string {
    const length = this.getRequiredParameter(parameters, 'length', 'MinLength');
    return `minLength(${length})`;
  }

  private static generateMaxLength(parameters?: Record<string, any>): string {
    const length = this.getRequiredParameter(parameters, 'length', 'MaxLength');
    return `maxLength(${length})`;
  }

  private static generateMatches(parameters?: Record<string, any>): string {
    const pattern = this.getRequiredParameter(parameters, 'pattern', 'Matches');
    // Escape backslashes for regex pattern in TypeScript
    const escapedPattern = String(pattern).replace(/\\/g, '\\\\');
    return `matches(new RegExp('${escapedPattern}'))`;
  }

  private static generateLessThan(parameters?: Record<string, any>): string {
    const value = this.getRequiredParameter(parameters, 'value', 'LessThan');
    return `lessThan(${value})`;
  }

  private static generateLessThanOrEqualTo(parameters?: Record<string, any>): string {
    const value = this.getRequiredParameter(parameters, 'value', 'LessThanOrEqualTo');
    return `lessThanOrEqualTo(${value})`;
  }

  private static generateGreaterThan(parameters?: Record<string, any>): string {
    const value = this.getRequiredParameter(parameters, 'value', 'GreaterThan');
    return `greaterThan(${value})`;
  }

  private static generateGreaterThanOrEqualTo(parameters?: Record<string, any>): string {
    const value = this.getRequiredParameter(parameters, 'value', 'GreaterThanOrEqualTo');
    return `greaterThanOrEqualTo(${value})`;
  }

  private static generateInclusiveBetween(parameters?: Record<string, any>): string {
    const min = this.getRequiredParameter(parameters, 'min', 'InclusiveBetween');
    const max = this.getRequiredParameter(parameters, 'max', 'InclusiveBetween');
    return `inclusiveBetween(${min}, ${max})`;
  }

  private static generateExclusiveBetween(parameters?: Record<string, any>): string {
    const min = this.getRequiredParameter(parameters, 'min', 'ExclusiveBetween');
    const max = this.getRequiredParameter(parameters, 'max', 'ExclusiveBetween');
    return `exclusiveBetween(${min}, ${max})`;
  }

  private static getRequiredParameter(
    parameters: Record<string, any> | undefined,
    paramName: string,
    validatorName: string
  ): any {
    if (!parameters || !(paramName in parameters)) {
      throw new Error(`Validator '${validatorName}' requires parameter '${paramName}'`);
    }
    return parameters[paramName];
  }

  private static formatValue(value: any): string {
    if (typeof value === 'string') {
      return `'${value}'`;
    } else if (typeof value === 'boolean') {
      return value.toString();
    } else {
      return String(value);
    }
  }
}

File: src/csharp-generator/Mapping/ValidatorMapping.cs
namespace FluentValidation.Generator.Mapping;

/// <summary>
/// Maps JSON validator names to C# FluentValidation method calls
/// </summary>
public static class ValidatorMapping
{
    /// <summary>
    /// Generates the C# FluentValidation method call for a given validator and its parameters
    /// </summary>
    /// <param name="validatorName">Name of the validator (e.g., "NotEmpty", "Length")</param>
    /// <param name="parameters">Optional parameters for the validator</param>
    /// <returns>C# method call string (e.g., "NotEmpty()", "Length(1, 100)")</returns>
    public static string MapToCSharp(string validatorName, Dictionary<string, object>? parameters)
    {
        return validatorName switch
        {
            "NotNull" => "NotNull()",
            "NotEmpty" => "NotEmpty()",
            "Empty" => "Empty()",
            "Null" => "Null()",

            "Equal" => GenerateEqual(parameters),
            "NotEqual" => GenerateNotEqual(parameters),

            "Length" => GenerateLength(parameters),
            "MinLength" => GenerateMinLength(parameters),
            "MaxLength" => GenerateMaxLength(parameters),

            "EmailAddress" => "EmailAddress()",
            "CreditCard" => "CreditCard()",

            "Matches" => GenerateMatches(parameters),

            "LessThan" => GenerateLessThan(parameters),
            "LessThanOrEqualTo" => GenerateLessThanOrEqualTo(parameters),
            "GreaterThan" => GenerateGreaterThan(parameters),
            "GreaterThanOrEqualTo" => GenerateGreaterThanOrEqualTo(parameters),

            "InclusiveBetween" => GenerateInclusiveBetween(parameters),
            "ExclusiveBetween" => GenerateExclusiveBetween(parameters),

            "IsInEnum" => "IsInEnum()",

            _ => throw new NotSupportedException($"Validator '{validatorName}' is not supported"),
        };
    }

    private static string GenerateEqual(Dictionary<string, object>? parameters)
    {
        var value = GetRequiredParameter(parameters, "value", "Equal");
        return $"Equal({FormatValue(value)})";
    }

    private static string GenerateNotEqual(Dictionary<string, object>? parameters)
    {
        var value = GetRequiredParameter(parameters, "value", "NotEqual");
        return $"NotEqual({FormatValue(value)})";
    }

    private static string GenerateLength(Dictionary<string, object>? parameters)
    {
        var min = GetRequiredParameter(parameters, "min", "Length");
        var max = GetRequiredParameter(parameters, "max", "Length");
        return $"Length({min}, {max})";
    }

    private static string GenerateMinLength(Dictionary<string, object>? parameters)
    {
        var length = GetRequiredParameter(parameters, "length", "MinLength");
        return $"MinimumLength({length})";
    }

    private static string GenerateMaxLength(Dictionary<string, object>? parameters)
    {
        var length = GetRequiredParameter(parameters, "length", "MaxLength");
        return $"MaximumLength({length})";
    }

    private static string GenerateMatches(Dictionary<string, object>? parameters)
    {
        var pattern = GetRequiredParameter(parameters, "pattern", "Matches");
        // Escape special characters in the pattern string
        var escapedPattern = pattern.ToString()?.Replace("\\", "\\\\").Replace("\"", "\\\"");
        return $"Matches(\"{escapedPattern}\")";
    }

    private static string GenerateLessThan(Dictionary<string, object>? parameters)
    {
        var value = GetRequiredParameter(parameters, "value", "LessThan");
        return $"LessThan({value})";
    }

    private static string GenerateLessThanOrEqualTo(Dictionary<string, object>? parameters)
    {
        var value = GetRequiredParameter(parameters, "value", "LessThanOrEqualTo");
        return $"LessThanOrEqualTo({value})";
    }

    private static string GenerateGreaterThan(Dictionary<string, object>? parameters)
    {
        var value = GetRequiredParameter(parameters, "value", "GreaterThan");
        return $"GreaterThan({value})";
    }

    private static string GenerateGreaterThanOrEqualTo(Dictionary<string, object>? parameters)
    {
        var value = GetRequiredParameter(parameters, "value", "GreaterThanOrEqualTo");
        return $"GreaterThanOrEqualTo({value})";
    }

    private static string GenerateInclusiveBetween(Dictionary<string, object>? parameters)
    {
        var min = GetRequiredParameter(parameters, "min", "InclusiveBetween");
        var max = GetRequiredParameter(parameters, "max", "InclusiveBetween");
        return $"InclusiveBetween({min}, {max})";
    }

    private static string GenerateExclusiveBetween(Dictionary<string, object>? parameters)
    {
        var min = GetRequiredParameter(parameters, "min", "ExclusiveBetween");
        var max = GetRequiredParameter(parameters, "max", "ExclusiveBetween");
        return $"ExclusiveBetween({min}, {max})";
    }

    private static object GetRequiredParameter(
        Dictionary<string, object>? parameters,
        string paramName,
        string validatorName
    )
    {
        if (parameters == null || !parameters.ContainsKey(paramName))
        {
            throw new InvalidOperationException(
                $"Validator '{validatorName}' requires parameter '{paramName}'"
            );
        }

        return parameters[paramName];
    }

    private static string FormatValue(object value)
    {
        return value switch
        {
            string s => $"\"{s}\"",
            bool b => b.ToString().ToLower(),
            _ => value.ToString() ?? "null",
        };
    }
}
